pkg/git-service/git_service.go

```diff
@@ -0,0 +1,111 @@
+package gitservice
+
+import (
+	"errors"
+	"regexp"
+	"strings"
+
+	"github.com/go-logr/logr"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type GitService struct {
+	GitURL      string
+	reference   string
+	secretValue string
+	gitType     GitProvider
+	owner       string
+	repo        string
+	logger      logr.Logger
+	status      metav1.ConditionStatus
+	reason      GitConditionReason
+}
+
+func New(gitURL, branch, secretValue string, logger logr.Logger) *GitService {
+	var status metav1.ConditionStatus = metav1.ConditionUnknown
+	var reason GitConditionReason = ReasonProcessing
+	gitType := identifyGitType(gitURL)
+	if gitType == Unknown {
+		logger.Error(errors.New(ReasonUnsupportedGitType.String()), "Unsupported Git Type")
+		return &GitService{
+			status: metav1.ConditionFalse,
+			reason: ReasonUnsupportedGitType,
+		}
+	}
+	owner, repo, err := getOwnerAndRepo(gitURL)
+	if err != nil {
+		logger.Error(err, "Cannot get owner and repo from gitURL")
+		return &GitService{
+			status: metav1.ConditionFalse,
+			reason: ReasonInvalidGitURL,
+		}
+	}
+
+	return &GitService{
+		GitURL:      gitURL,
+		reference:   branch,
+		gitType:     gitType,
+		owner:       owner,
+		repo:        repo,
+		secretValue: secretValue,
+		logger:      logger,
+		status:      status,
+		reason:      reason,
+	}
+}
+
+func (g *GitService) IsRepoReachable() (metav1.ConditionStatus, GitConditionReason) {
+	if g.status != metav1.ConditionUnknown {
+		return g.status, g.reason
+	}
+	switch g.gitType {
+	case Github:
+		isGHRepoReachable(g)
+	case Gitlab:
+		isGLRepoReachable(g)
+	}
+	return g.status, g.reason
+}
+
+func identifyGitType(gitURL string) GitProvider {
+	switch {
+	case strings.Contains(gitURL, "github.com"):
+		return Github
+	case strings.Contains(gitURL, "gitlab.com"):
+		return Gitlab
+	default:
+		return Unknown
+	}
+}
+
+func getOwnerAndRepo(gitURL string) (string, string, error) {
```

Comment:
not super famous but seems to be doing our job : https://github.com/kubescape/go-git-url

Something to look into.

Reply:
Yeah, the codebase is very neat. A lot of patterns to learn from it.
Maybe I can use it for the Bitbucket and Azure implementations.